pkg/archive/copy.go

```diff
@@ -0,0 +1,405 @@
+		// extracted to the parent directory. Use the destination basename.
+		dstDir, dstBase = filepath.Split(dstPath)
+	default:
+		return err
+	}
+
+	// We may need to swap the Name of the entry to match the dstPath so an
+	// altered tar archive will need to be constructed. We also want to ensure
+	// that we extract exactly one entry.
+	r, w := io.Pipe()
+	errC := promise.Go(func() error {
+		defer w.Close()
+
+		tarReader := tar.NewReader(content)
+		hdr, err := tarReader.Next()
+		if err != nil {
+			return err
+		}
+
+		if hdr.Typeflag == tar.TypeDir || filepath.Dir(hdr.Name) != "." {
+			return ErrCannotCopyDir
+		}
+
+		// Alter the basename of the file if specified.
+		if dstBase != "" {
+			hdr.Name = dstBase
+		}
+
+		tarWriter := tar.NewWriter(w)
+		defer tarWriter.Close()
+
+		if err = tarWriter.WriteHeader(hdr); err != nil {
+			return err
+		}
+
+		// Fully copy only the first entry to the tar writer.
+		_, err = io.Copy(tarWriter, tarReader)
+
+		return err
+	})
+
+	defer func() {
+		if er := <-errC; err != nil {
+			err = er
+		}
+	}()
+
+	log.Debugf("copying file to %q", dstDir)
+
+	return archiver.Untar(r, dstDir, options)
+}
+
+// copyDirTo handles copying the directory from the given archive to  the
+// destination path. content should have been confirmed to be a directory by
+// the outer call to CopyTo(). When unpacking, contents are filtered to only
+// the directory which the given wrapper detected from the content archive. The
+// directory and its contents are copied to the specified dstPath preserving
+// all metadata.
+//
+// If dstPath does not exist, when unpacking, the dirname is replaced by the
+// basename of dstPath, and the archive is extracted into dstPath's parent
+// directory. It is an error if dstPath exists and is not a directory.
+func copyDirTo(archiver *Archiver, content *copiedContentArchiveWrapper, dstPath string, options *TarOptions) error {
+	log.Debugf("copying directory to %q", dstPath)
+
+	// Determine if dstPath exists and is a directory.
+	// The error is not nil if the destination does not exist, if it exists as
+	// a file *but* dstPath is asserted to be a directory, or any other stat
+	// error.
+	dstStat, err := os.Stat(dstPath)
+
+	var dstDir string
+
+	switch {
+	case err == nil && dstStat.IsDir():
+		// Exists and is a directory.
+		dstDir = dstPath
+	case err == nil:
+		// Exists and is a file.
+		return ErrNotDirectory
+	case os.IsNotExist(err):
+		// The parent directory is confirmed to exist by the outer call to
+		// CopyTo(). The destination directory will be created when the archive
+		// is extracted to the parent directory.
+		dstDir = filepath.Dir(filepath.Clean(dstPath))
+		options.Replace = [2]string{content.dirname, filepath.Base(filepath.Clean(dstPath))}
+	default:
+		return err
+	}
+
+	options.IncludeFiles = []string{content.dirname + "/"}
+
+	log.Infof("copying dir to %q: filter: %v, replace: %v", dstDir, options.IncludeFiles, options.Replace)
```

Comment:
Shouldn't this be debug? To me this seems too implementation specific/cryptic to print out on every execution. Silence means success.